src/jsonrpc.c

```diff
@@ -0,0 +1,493 @@
+/*
+ * Copyright (c) 2016 Raphaël Monrouzeau <raphael.monrouzeau@gmail.com>
+ *
+ * Permission to use, copy, modify, and distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#include <stdbool.h>
+
+#include <yajl/yajl_tree.h>
+#include <yajl/yajl_gen.h>
+
+#include "kore.h"
+#include "http.h"
+#include "jsonrpc.h"
+
+static void
+init_log(struct jsonrpc_log *log)
+{
+	log->msg = NULL;
+	log->next = log;
+	log->prev = log;
+}
+
+static void
+append_log(struct jsonrpc_log *prev, int lvl, char *msg)
+{
+	struct jsonrpc_log *new = kore_malloc(sizeof(struct jsonrpc_log));
+
+	new->lvl = lvl;
+	new->msg = msg;
+
+	new->prev = prev;
+	new->next = prev->next;
+	prev->next->prev = new;
+	prev->next = new;
+}
+
+static void
+free_log(struct jsonrpc_log *root)
+{
+	for (struct jsonrpc_log *it = root->next; it != root; it = it->next) {
+		kore_mem_free(it);
+	}
+}
+
+static void
+init_request(struct jsonrpc_request *req)
+{
+	init_log(&req->log);
+	kore_buf_init(&req->buf, 256);
+	req->gen = NULL;
+	req->http = NULL;
+	req->json = NULL;
+	req->id = NULL;
+	req->method = NULL;
+	req->params = NULL;
+	req->log_levels = (1 << LOG_EMERG) | (1 << LOG_ERR) | (1 << LOG_WARNING)
+			| (1 << LOG_NOTICE);
+        req->flags = 0;
+}
+
+static void
+free_request(struct jsonrpc_request *req)
+{
+	if (req->gen != NULL) {
+		yajl_gen_free(req->gen);
+		req->gen = NULL;
+	}
+	if (req->json != NULL) {
+		yajl_tree_free(req->json);
+		req->json = NULL;
+	}
+	kore_buf_destroy(&req->buf);
+	free_log(&req->log);
+}
+
+void
+jsonrpc_log(struct jsonrpc_request *req, int lvl, const char *fmt, ...)
+{
+	va_list		ap;
+	char		*msg;
+	u_int32_t	start = req->buf.offset;
+
+	va_start(ap, fmt);
+	kore_buf_appendv(&req->buf, fmt, ap);
+	va_end(ap);
+
+	msg = kore_buf_stringify(&req->buf, NULL) + start;
+
+	append_log(&req->log, lvl, msg);
+}
+
+static int
+read_json_body(struct http_request *http_req, ssize_t body_max_len,
```

Comment:
I guess I wanted a more fine grained policy mechanism, but anyhow it would necessitate stream parsing the JSON and could be implemented by the application. I removed it and will try to amend my pull request (I'm quite new to the pull request system so this may take some days).